### PR TITLE
Add cache to ParamRef.underlying

### PR DESCRIPTION
**Pull request by nicolasstucki**

### CODE REVIEW

**Comment by nicolasstucki** [comment]
test performance please

**Comment by nicolasstucki** [comment]
test performance please

File: compiler/src/dotty/tools/dotc/core/Types.scala [hunk 1]
```diff
@@ -3187,10 +3187,16 @@ object Types {
     def paramNum: Int
     def paramName: binder.ThisName = binder.paramNames(paramNum)
 
+    private[this] var cachedUnderlying: Type = null
     override def underlying(implicit ctx: Context): Type = {
-      val infos = binder.paramInfos
-      if (infos == null) NoType // this can happen if the referenced generic type is not initialized yet
-      else infos(paramNum)
+      if (cachedUnderlying ne null) {
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/core/Types.scala]: This should be eq not ne (or just == since it will be translated to the same bytecode)

**Comment by nicolasstucki** [comment]
test performance please

**Comment by nicolasstucki** [comment]
test performance please

**Comment by nicolasstucki** [comment]
test performance please

**Comment by nicolasstucki** [comment]
test performance please

**Pull request closed without merging**